karapace/serialization.py

```diff
@@ -207,8 +262,15 @@ def read_value(schema: TypedSchema, bio: io.BytesIO):
 
 def write_value(schema: TypedSchema, bio: io.BytesIO, value: dict) -> None:
     if schema.schema_type is SchemaType.AVRO:
+
+        # Backwards compatibility: Support JSON encoded data without the tags for unions.
+        if avro.io.Validate(schema.schema, value):
+            data = value
+        else:
+            data = flatten_unions(schema.schema, value)
```

Comment:
This is now more forgiving than it needs to be, because the old format is handled in the branch above. A valid approach here would be to reuse the old version of flatten that has strict validation

https://github.com/aiven/karapace/pull/285